pkg/virt-controller/watch/migration.go

```diff
@@ -49,84 +70,183 @@ func (md *MigrationDispatch) Execute(store cache.Store, queue workqueue.RateLimi
-			if err != nil {
-				logger.Error().Reason(err).Msg("could not copy migration object")
-				queue.AddRateLimited(key)
+	var migration *v1.Migration = obj.(*v1.Migration)
+	logger := logging.DefaultLogger().Object(migration)
+
+	vm, exists, err := md.vmService.FetchVM(migration.Spec.Selector.Name)
+	if err != nil {
+		logger.Error().Reason(err).Msgf("fetching the vm %s failed", migration.Spec.Selector.Name)
+		queue.AddRateLimited(key)
+		return
+	}
+
+	if !exists {
+		logger.Info().Msgf("VM with name %s does not exist, marking migration as failed", migration.Spec.Selector.Name)
+		setMigrationFailed(migration)
+		return
+	}
+
+	switch migration.Status.Phase {
+	case v1.MigrationUnknown:
+		// Fetch vm which we want to migrate
+
+		if vm.Status.Phase != v1.Running {
+			logger.Error().Msgf("VM with name %s is in state %s, no migration possible. Marking migration as failed", vm.GetObjectMeta().GetName(), vm.Status.Phase)
+			setMigrationFailed(migration)
+			return
+		}
+
+		if err := mergeConstraints(migration, vm); err != nil {
+			logger.Error().Reason(err).Msg("merging Migration and VM placement constraints failed.")
+			queue.AddRateLimited(key)
+			return
+		}
+		podList, err := md.vmService.GetRunningVMPods(vm)
+		if err != nil {
+			logger.Error().Reason(err).Msg("could not fetch a list of running VM target pods")
+			queue.AddRateLimited(key)
+			return
+		}
+
+		numOfPods, targetPod := investigateTargetPodSituation(migration, podList)
+
+		if targetPod == nil {
+			if numOfPods > 1 {
+				logger.Error().Reason(err).Msg("another migration seems to be in progress, marking Migration as failed")
+				// Another migration is currently going on
+				setMigrationFailed(migration)
 				return
+			} else if numOfPods == 1 {
+				// We need to start a migration target pod
+				// TODO, this detection is not optimal, it can lead to strange situations
+				err := md.vmService.CreateMigrationTargetPod(migration, vm)
+				if err != nil {
+					logger.Error().Reason(err).Msg("creating a migration target pod failed")
+					queue.AddRateLimited(key)
+					return
+				}
+
 			}
-			// Fetch vm which we want to migrate
-			vm, exists, err := md.vmService.FetchVM(migration.Spec.Selector.Name)
-			if err != nil {
-				logger.Error().Reason(err).Msgf("fetching the vm %s failed", migration.Spec.Selector.Name)
+		} else {
+			// Unlikely to hit this case, but prevents erroring out
+			// if we re-enter this loop
+			logger.Info().Msg("Migration appears to be set up, but was not set to InProgress.")
+		}
+		setMigrationPhase(migration, v1.MigrationInProgress)
+	case v1.MigrationInProgress:
+		podList, err := md.vmService.GetRunningVMPods(vm)
+		if err != nil {
+			logger.Error().Reason(err).Msg("could not fetch a list of running VM target pods")
+			queue.AddRateLimited(key)
+			return
+		}
+
+		_, targetPod := investigateTargetPodSituation(migration, podList)
+
+		if targetPod == nil {
+			setMigrationFailed(migration)
```

Comment:
In the case of a final return, we have to always do a `queue.Forget(key)`. Otherwise, if a migration with the same name is posted again, it might be processed delayed. Also the controller would see a fail history ...